provisioner/data.js

```diff
@@ -1,30 +1,615 @@
+    'Placement',
+    'RamdiskId',
+    'SubnetId',
+  ]);
+
+  // Now check that there are no unknown keys
+  Object.keys(launchSpec).forEach(function(key) {
+    assert(allowedKeys.indexOf(key) !== -1, 'Your launch spec has invalid key ' + key);
+  });
+
+  // These are keys which we do not allow in the generated launch spec
+  var disallowedKeys = [
+    'Placement',
+  ];
+
+  disallowedKeys.forEach(function(key) {
+    assert(!launchSpec[key], 'Your launch spec must not have key ' + key);
+  }); 
+
+  return launchSpec;
+};
+
+
+/**
+ * Make sure that all combinations of LaunchSpecs work.  This sync
+ * function will throw if there is an error found or will return
+ * a dictionary of all the launch specs!
+ */
+WorkerType.testLaunchSpecs = function(worker, keyPrefix, provisionerId) {
+  assert(worker);
+  assert(keyPrefix);
+  assert(provisionerId);
+  var errors = [];
+  var launchSpecs = {};
+  Object.keys(worker.regions).forEach(function(region) {
+    launchSpecs[region] = {};
+    Object.keys(worker.types).forEach(function(type) {
+      try {
+        var x = WorkerType.createLaunchSpec(region, type, worker, keyPrefix, provisionerId);
+        launchSpecs[region][type] = x;
+      } catch (e) {
+        errors.push(e)
+      }
+    });
+  });
+  if (errors.length > 0) {
+    var err = new Error('Launch specifications are invalid');
+    err.code = 'InvalidLaunchSpecifications';
+    err.reasons = errors;
+    throw err;
+  }
+  return launchSpecs;
+};
+
+
+/**
+ * Figure out how many capacity units need to be created.  This number is
+ * determined by calculating how much capacity is needed to maintain a given
+ * scaling ratio and returns the number of capacity units which need to be
+ * created or destroyed.  This will give an exact number of units, something
+ * else will be required to decide what to do if the number of needed capacity
+ * units does not fit nicely with the number of capacity units available per
+ * instance type.  Positive value means add capacity, negative means destroy.
+ * Running capacity are nodes that are on and assumed to be accepting jobs.
+ * Since we don't currently have a way to know when the node is *actually*
+ * accepting jobs, we'll just assume that running == accepting jobs.
+ * Pending capacity are those units (pending instances, spot requests)
+ * which are going to be created but have not yet been created.  We want to
+ * offset the number of units that we'd be creating by the number of pending
+ * capacity units
+ */
+WorkerType.prototype.determineCapacityChange = function(runningCapacity, pendingCapacity, pending) {
+  assert(typeof runningCapacity === 'number');
+  assert(typeof pendingCapacity === 'number');
+  assert(typeof pending === 'number');
+
+  // We need to know how many total capacity units are extant
+  var totalCapacity = pendingCapacity + runningCapacity;
+
+  // We need to know the current ratio of capacity to pending
+  var percentPending = 1 + pending / totalCapacity;
```

Comment:
I'm not sure what the `1 + ` is there for... if you're trying to avoid `0/0` you need parentheses...

I suspect we need to do this differently... We need to support `scalingRatio = 0` meaning that when there is more than 0% pending tasks (ie any pending tasks) we should scale up.